### PR TITLE
Small cleanups, add a `DataTree` pretty printer

### PR DESCRIPTION
**Pull request by tybug**
No user-facing changes, just (1) some cleanups, (2) changes in preparation for the shrinker ir migration, and (3) a pretty printer for `DataTree`.

### CODE REVIEW

File: hypothesis-python/tests/conjecture/test_data_tree.py [hunk 1]
```diff
@@ -567,3 +568,10 @@ def buf(data):
         prefix = tree.generate_novel_prefix(Random())
         data = ConjectureData.for_buffer(prefix)
         assert data.draw_float(min_value, max_value, allow_nan=False) == expected_value
+
+
+def test_datatree_repr():
+    tree = DataTree()
+    observer = tree.new_observer()
+    observer.draw_boolean(True, was_forced=False, kwargs={"p": 0.5})
+    assert pretty.pretty(tree) == "boolean True {'p': 0.5}"
```
Comment (Zac-HD) [inline hypothesis-python/tests/conjecture/test_data_tree.py:577]: Let's test some more complex trees too

File: hypothesis-python/tests/cover/test_simple_collections.py [hunk 2]
```diff
@@ -113,9 +113,7 @@ def test_sets_of_fixed_length(n):
 @pytest.mark.parametrize("n", range(10))
 def test_dictionaries_of_fixed_length(n):
     x = set(
-        minimal(
-            dictionaries(integers(), booleans(), min_size=n, max_size=n), lambda x: True
-        ).keys()
+        minimal(dictionaries(integers(), booleans(), min_size=n, max_size=n)).keys()
```
Comment (Zac-HD) [inline hypothesis-python/tests/cover/test_simple_collections.py:116]:
```suggestion
        minimal(dictionaries(integers(), booleans(), min_size=n, max_size=n))
```

File: hypothesis-python/tests/cover/test_simple_collections.py [hunk 3]
```diff
@@ -168,9 +166,10 @@ def test_small_sized_sets(x):
 
 
 def test_minimize_dicts_with_incompatible_keys():
-    assert minimal(
-        fixed_dictionaries({1: booleans(), "hi": lists(booleans())}), lambda x: True
-    ) == {1: False, "hi": []}
+    assert minimal(fixed_dictionaries({1: booleans(), "hi": lists(booleans())})) == {
+        1: False,
+        "hi": [],
+    }
```
Comment (Zac-HD) [inline hypothesis-python/tests/cover/test_simple_collections.py:172]:
```suggestion
    strat = fixed_dictionaries({1: booleans(), "hi": lists(booleans())})
    assert minimal(strat) == {1: False, "hi": []}
```

File: hypothesis-python/tests/conjecture/test_data_tree.py (hunk 1 above)
Comment (tybug) [reply hypothesis-python/tests/conjecture/test_data_tree.py]: I added a more complicated tree. I can go all out testing this even more if you want, but I wasn't sure how useful testing an effectively debug-only feature would be.

File: hypothesis-python/tests/conjecture/test_data_tree.py [hunk 4]
```diff
@@ -567,3 +570,44 @@ def buf(data):
...
+        == textwrap.dedent(
+            f"""
+        boolean True {bool_kwargs}
+          Conclusion (Status.INVALID)
+        boolean False {bool_kwargs}
+          integer 42 {int_kwargs}
+            Conclusion (Status.VALID)
+          integer 0 {int_kwargs}
+            boolean False [forced] {bool_kwargs}
+              Conclusion (Status.INTERESTING, {origin})
```
Comment (Zac-HD) [inline hypothesis-python/tests/conjecture/test_data_tree.py:610]: Nice!  Ideas for future: 
- should we sort the child branches to show shrink order?
- explicitly mark exhausted subtrees?

**Review by Zac-HD — APPROVED** [review]

**Pull request merged**